File: Week_01/Practicum/python/Group the People Given the Group Size They Belong To.py
class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        
        count_lst = [[] for _ in range(501)]

        for i, num in enumerate(groupSizes):
            count_lst[num].append(i)
        
        groups = []

        for idx, group_lst in enumerate(count_lst):
            if len(group_lst) == 0:
                continue

            groups_count = int(len(group_lst) / idx)
            group_idx = 0
            for group in range(groups_count):
                current_group = []
                for _ in range(idx):
                    current_group.append(group_lst[group_idx])
                    group_idx += 1

                groups.append(current_group)

        return groups

File: Week_01/Practicum/К/Group the People Given the Group Size They Belong To.cpp
// За решението може да използваме един помощен масив от вектори groups, в който
// на позиция i се намира група от i на брой души. 
// Обхождаме вектора groupSizes, като всеки един човек поставяме в група от groupSizes[i] на брой души. 
// Когато размерът на групата стане groupSizes[i], тогава я добавяме в отговора (вектора grouped)
// и изтриваме (group.clear()) хората от нея, защото вече са разпределени (така изполваме вектора groups[i] за всички групи от i на брой души)
class Solution {
    vector<int> groups[501] {};
public:
    vector<vector<int>> groupThePeople(vector<int>& groupSizes) {
        vector<vector<int>> grouped;
        for (int i = 0; i < groupSizes.size(); ++i) {
            auto& group = groups[groupSizes[i]];
            group.push_back(i);

            if (group.size() == groupSizes[i]) {
                grouped.push_back(group);
                group.clear();
            }

        }
        return grouped;
    }
};